cpu/fe310/thread_arch.c

```diff
@@ -180,8 +180,10 @@ void thread_yield_higher(void)
     /* Use SW intr to schedule context switch */
     CLINT_REG(CLINT_MSIP) = 1;
 
-    /* Latency of SW intr can be 4-7 cycles; wait for the SW intr */
-    __asm__ volatile ("wfi");
+    /* Latency of SW intr can be a few cycles; wait for the SW intr.
+     * We cannot use WFI here as the SW intr may be delivered before we
+     * reach the WFI instruction, thereby causing a thread deadlock. */
+    while (CLINT_REG(CLINT_MSIP) == 1) {};
```

Comment:
If anyone has an idea for a better loop condition let me know. Please also double check that this condition isn't subject to race conditions.

Reply:
this seems reasonable. IIRC, previously we just had 7 nops...

Reply:
How about
```C
do {
    __asm__ volatile ("wfi");
} while (CLINT_REG(CLINT_MSIP) == 1)
```
?

Reply:
How does that prevent the interrupt from being delivered before the WFI is reached?

Reply:
You're right, I didn't consider that case. Nothing to add then.

Reply:
So ACK?

Reply:
Ping.